### PR TITLE
KFLUXINFRA-1593: unit tests for common.go functions and backend_probe.go -> CheckAvailability

### PR DESCRIPTION
**Pull request by meyrevived**
This pull request contains test suites for backend_probe.go -> CheckAvailability (as the Jira task was meant to be about) but also contains a separate test suite for common.go's functions which handle common.go's probes map for managing the successes and failures of platforms. 
This PR also includes a fix for the race conditions found in the tested common.go functions, proven by the test to cause a [fatal error](https://drive.google.com/file/d/1hz1Ok3MMnXDQbsaqyhuy29wyNHnP0R61/view?usp=sharing).

**CheckAvailability test suite:**
This test suite covers the only testable function in backend_probe.go - CheckAvailability.  Is split into two main sections:

1. Testing the logic of returning nothing (successful platform) or returning an error (failing platform). This is not only tested by creating conditions for a successful/failing platform, but also a boundary test for the errorThreshold constant that dynamically evaluates the boundaries to test for future robustness.

2. Test the state change of a platform as CheckAvailability is called twice, verifying the Swap() action clears all successes and failures of a probe.

**Race conditions in common.go functions test suite:**
A unit test for the possibility of race conditions hitting the functions in common.go that handle the probes map.
 This test is designed to reliably trigger race conditions via goroutines as persistent readers and writers to the probes map, running for a controlled duration.
This test's meaning of `passing` is `nothing panicked because of race conditions on probes`.

Assisted-by: Gemini

### CODE REVIEW

**Review by mshaposhnik — APPROVED** [review]

**Review by sadlerap — DISMISSED** [review]

File: pkg/metrics/common.go [hunk 1]
```diff
@@ -14,6 +15,7 @@ const (
 
 var (
 	probes                      = map[string]*AvailabilityProbe{}
+	probesMutex                 = &sync.RWMutex{} // Read-Write mutex for probes
```
[the remark this answers is not in the document]
Comment (sadlerap) [reply pkg/metrics/common.go:18]: We shouldn't need a lock at all here, since we only obtain a writer lock for initializing the availability probe entry.  Once this is done, we read the entry and call `Success()` or `Failure()`, which is always safe to do and doesn't require synchronization.

If we wanted to use a RWMutex, the `CountAvailabilitySuccess`/CountAvailabilityError` functions should only need to take read locks, and they'd upgrade to a writer lock only if the platform's availability probes haven't been initialized yet.

Ideally, we'd clean up this interface so that we obtain a list of platforms and pre-initialize everything before we start tracking metrics, but that might be outside the scope of this PR.

File: pkg/metrics/common_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	// This test's meaning of `passing` is `nothing panicked because of race conditions on probes`.
+	Describe("common.go race condition test", func() {
+
+		Context("when the probes map is accessed concurrently", func() {
+			BeforeEach(func() {
+				probes = make(map[string]*AvailabilityProbe)
+			})
+
+			It("should not have a data race", func() {
+				ctx, cancel := context.WithCancel(context.Background())
```
Comment (filariow) [inline pkg/metrics/common_test.go:27]:
```suggestion
			It("should not have a data race", func(ctx context.Context) {
				ctx, cancel := context.WithCancel(ctx)
```

File: pkg/metrics/common_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+							case <-ctx.Done():
+								return
+							default:
+								randomPlatform := fmt.Sprintf("platform-%d", rand.Intn(100))
+								CountAvailabilitySuccess(randomPlatform)
+								CountAvailabilityError(randomPlatform)
+							}
+						}
+					}()
+				}
```
Comment (filariow) [inline pkg/metrics/common_test.go:65]: may you add some comments here? I find it a bit hard to follow

File: pkg/metrics/common.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (meyrevived) [reply pkg/metrics/common.go:18]: OK, @sadlerap, for the sake of testing, I applied you proposed change to the current use of the RWMutex like so:
```
func CountAvailabilitySuccess(platform string) {

	if probes[platform] == nil {
		probesMutex.Lock() // Use a full lock for exclusive write access to guard against race conditions
		defer probesMutex.Unlock()

		watcher := NewBackendProbe()
		probes[platform] = &watcher
	}
	(*probes[platform]).Success()
}

func CountAvailabilityError(platform string) {

	if probes[platform] == nil {
		probesMutex.Lock() // Same exclusive lock as in CountAvailabilitySuccess
		defer probesMutex.Unlock()
		
		watcher := NewBackendProbe()
		probes[platform] = &watcher
	}
	(*probes[platform]).Failure()
}
```

This [still fails the race conditions test](https://drive.google.com/file/d/1OhbfTjNtwPihP8_PHbC7eWWg3ve4IzQt/view?usp=sharing)

We can do a double-check and use the write lock more carefully. It'll mean way, way, way more code:
```
func CountAvailabilitySuccess(platform string) {

	// Gotta use a read lock to check if the probe exists, but release it ASAP
	probesMutex.RLock()
	probe, ok := probes[platform]
	probesMutex.RUnlock()

	if ok {
		// Probe exists, no write lock needed.
		(*probe).Success()
		return
	}

	// The probe doesn't exist, take a full write lock.
	probesMutex.Lock()
	defer probesMutex.Unlock()

	// We MUST check again, because another goroutine might have created it in the tiny gap between our RUnlock() and
	// Lock() calls.
	if probe, ok = probes[platform]; !ok {
		// It still doesn't exist, so we are the one to create it.
		newProbe := NewBackendProbe()
		probes[platform] = &newProbe
		probe = &newProbe
	}

	(*probe).Success()
}

func CountAvailabilityError(platform string) {

	// Same story here
	probesMutex.RLock()
	probe, ok := probes[platform]
	probesMutex.RUnlock()

	if ok {
		(*probe).Failure()
		return
	}

	probesMutex.Lock()
	defer probesMutex.Unlock()

	if probe, ok = probes[platform]; !ok {
		newProbe := NewBackendProbe()
		probes[platform] = &newProbe
		probe = &newProbe
	}

	(*probe).Failure()
}
```

Granted, it doesn't break the race condition test, but I suspect it'll give @filariow a rash seeing so many more code lines :smiley: 

sync.Map can be used, it's slightly less type-safe so we'll need type-assertion (a.k.a, more code lines) like so:
```
var (

	probes = sync.Map{}

	// unchanged code
)

// unchanged code 
func checkProbes(ctx context.Context) {
	
	log := logr.FromContextOrDiscard(ctx)
	
	probes.Range(func(key, value any) bool {
		// Type assertions are required for both key and value.
		platform := key.(string)
		probe := value.(*AvailabilityProbe)

		checkLabel := prometheus.Labels{"check": platform}
		checkErr := (*probe).CheckAvailability(ctx)
		if checkErr != nil {
			log.Error(checkErr, "Failing availability probe due to high error threshold", "platform", platform)
			controllerAvailabilityGauge.With(checkLabel).Set(0)
		} else {
			log.Info("Availability probe OK", "platform", platform)
			controllerAvailabilityGauge.With(checkLabel).Set(1)
		}
		// Return true to continue iterating over the map.
		return true
	})
}

func CountAvailabilitySuccess(platform string) {
	// Create a new probe instance to be stored if the key is missing.
	newProbe := NewBackendProbe()
	// Either return the existing value that was found, or a new value which is then stored.
	actual, _ := probes.LoadOrStore(platform, &newProbe)

	// Gotta re-type-assert because the returned value is of type 'any'.
	probe := actual.(*AvailabilityProbe)
	(*probe).Success()

}

func CountAvailabilityError(platform string) {
	// Exact same story here.
	newProbe := NewBackendProbe()

	actual, _ := probes.LoadOrStore(platform, &newProbe)

	probe := actual.(*AvailabilityProbe)
	(*probe).Failure()
}
```
`sync.Map` doesn't break the race condition test either, so as far as I'm concerned, I have no issue with either alternative. I leave it up to the three of you to discuss @mshaposhnik, @filariow and @sadlerap

File: pkg/metrics/backend_probe_test.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		ctx   context.Context
+	)
+
+	BeforeEach(func() {
+		ctx = context.TODO()
+	})
+
+	// Testing the logic determining whether a platform is successful or failing
+	When("the probe is healthy", func() {
+		It("should pass with no successes and no failures", func() {
```
[the remark this answers is not in the document]
Comment (meyrevived) [reply pkg/metrics/backend_probe_test.go:30]: Will be implemented in the next commit

File: pkg/metrics/backend_probe_test.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+			probe = setupProbeWithCounts(0, 0)
+			err := probe.CheckAvailability(ctx)
+			Expect(err).ShouldNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (meyrevived) [reply pkg/metrics/backend_probe_test.go:33]: Will be implemented in the next commit

File: pkg/metrics/common_test.go (hunk 2 above)
Comment (meyrevived) [reply pkg/metrics/common_test.go:27]: Cool! Will be implemented in the next commit

File: pkg/metrics/common_test.go (hunk 3 above)
Comment (meyrevived) [reply pkg/metrics/common_test.go:65]: Sure, check out the code in the corrections and fixes commit.

File: pkg/metrics/common_test.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+				time.Sleep(200 * time.Millisecond)
+
+				cancel()
```
[the remark this answers is not in the document]
Comment (meyrevived) [reply pkg/metrics/common_test.go:69]: Hopefully the comments I've added to the fixes commit will explain this, but in general,  `defer cancel() `is scheduled. It won't run until the entire `It()` block is about to return. The `cancel()` inside the `It()` block allows the goroutine writers to slowly wind down as they all `wg.Wait() ` for all of them to get to  `wg.Done() ` . With no internal `cancel()` this `wg.Done()` does not happen - each writer is stuck in the for loop and we never get to the `defer cancel()` --> we're in a deadlock.

File: pkg/metrics/common_test.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+				// Allow the concurrent goroutine writers to run for a set duration, creating the necessary window for
+				// a data race to occur between themselves.
+				time.Sleep(200 * time.Millisecond)
+
+				// Explicitly signals all goroutine writers to stop their work. This is critical for a graceful shutdown
+				// before we wait for them so that the wg.Wait() call can complete.
+				cancel()
+				// A synchronization point. It blocks until all goroutines have received the cancellation signal and
+				// called wg.Done().
+				wg.Wait()
```
Comment (filariow) [inline pkg/metrics/common_test.go:78]: thanks for adding the comments, now it's way better. The test mostly looks good to me. The only thing I'm not super convinced of is the `time.Sleep(200*time.Millisecond)`. The number of operations performed in that timeframe will depend on the machine and on the load. What if, instead of waiting, we count the number of operations? Something like the following:

```suggestion
			It("should not have a data race", func(tctx context.Context) {
				ctx, cancel := context.WithCancel(tctx)
				// This cancel() is a safety net to ensures that if the test ever panics or returns early for any
				// possible reason, the context is always guaranteed to be cleaned up and prevent goroutine leaks.
				defer cancel()

				var wg sync.WaitGroup
				writerCount := 5
				writesPerWriter := 10000

				// Persistent reader
				wg.Add(1)
				go func() {
					defer wg.Done()
					defer GinkgoRecover()
					for range writerCount * writesPerWriter {
						checkProbes(ctx)
					}
				}()

				// Persistent writers - these writers continuously attempt to add new entries to probes, which is the
				// "write" part of the race condition created in the test.
				for i := 0; i < writerCount; i++ {
					wg.Add(1)
					go func() {
						defer wg.Done()
						defer GinkgoRecover()
						for range writesPerWriter {
							randomPlatform := fmt.Sprintf("platform-%d", rand.Intn(100))
							CountAvailabilitySuccess(randomPlatform)
							CountAvailabilityError(randomPlatform)
						}
					}()
				}
				wg.Wait()
```

File: pkg/metrics/common_test.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	// to the probes map, running for a controlled duration.
+	// This test's meaning of `passing` is `nothing panicked because of race conditions on probes`.
+	Describe("common.go race condition test", func() {
+
+		Context("when the probes map is accessed concurrently", func() {
+			BeforeEach(func() {
+				probes = make(map[string]*AvailabilityProbe)
+			})
+
+			It("should not have a data race", func(tctx context.Context) {
```
Comment (filariow) [inline pkg/metrics/common_test.go:26]: we don't have any guarantee writers and readers are acting at the same time. Ideas on how to make sure this is happening? checking value in `controllerAvailabilityGauge` maybe?

File: pkg/metrics/common.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadlerap) [reply pkg/metrics/common.go:18]: I don't think that's going to happen in practice, because we only take a writer lock to initialize a probe manager for a platform we've not seen yet.  Once it exists, we only need to take a read lock, and in the limit, we only take read locks.

This is why I was considering using a `sync.Map` - if you're only doing reads, you don't need to take any internal locks.  IMO our access pattern fits a `sync.Map` better.

An alternative idea I had: we could pre-initialize `probes` with all the platforms we see in the config and leave it read-only.  Since `AvailabilityProbes` are already safe to access from multiple threads, we wouldn't need any kind of locking on the map.  However, this might be a bit much for the scope of what's already in this PR.

File: pkg/metrics/common_test.go (hunk 8 above)
Comment (meyrevived) [reply pkg/metrics/common_test.go:26]: I have full faith in Ginkgo's and Golang's race detector to make sure writers an readers ran concurrently. If we go about engineering prometheus and involve some timing-dependent logic to make sure a reader set `controllerAvailabilityGauge`'s gauge to 1 and then a writer did something and then the reader re-set it to 0, that would really be an overkill instead of something golang provided for us out of the box.

File: pkg/metrics/common.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadlerap) [reply pkg/metrics/common.go:18]: > That is a very good point. Also, since you're the last reviewer still commenting on this debate - your vote (and, technically, what Max also wanted) is the winner -> the fix stays as it is for this PR. @sadlerap if you feel a larger fix outside the scope of this is needed, I'm sure it's a great idea and would benefit MPC as an independent PR.

I think I'll file my proposal as something to do in a follow-up PR.  (Github helpfully didn't show me your comment until after I posted the above comment :cry:)
